File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let start_i=0,end_i=nums.length-1
    
    while(start_i<=end_i){
        let mid=Math.floor(((start_i+end_i)/2))
        if(nums[mid]==target )
            return mid;
        
        if(nums[start_i] <= nums[mid]){
            if(target >= nums[start_i] && target<nums[mid] ){
                end_i=mid-1 
            }
            else{
                start_i=mid+1 
            }
        }
        else{
            if(target>nums[mid] && target<=nums[end_i]){
                start_i=mid+1
            }
            else{
                end_i=mid-1
            }
        }
    }
        
    return  -1; 
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
    start_i, end_i = 0, nums.length - 1
    
    while start_i <= end_i
        mid = (start_i + end_i) / 2
        
        return mid if nums[mid] == target
        
        if nums[start_i] <= nums[mid]
            if target >= nums[start_i] && target < nums[mid]
                end_i = mid - 1
            else
                start_i = mid + 1
            end
        else
            if target > nums[mid] && target <= nums[end_i]
                start_i = mid + 1
            else
                end_i = mid - 1
            end
        end
    end
    
    return -1
end